### PR TITLE
Add EF core implementation

### PR DESCRIPTION
**Pull request by TheDioBrando**
Add DLL ConsoleManagement

### CODE REVIEW

File: HW4/Models/Books.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿namespace HW4.Models
+{
+    public class DbBooks
```
Comment (vorobevamv) [inline HW4/Models/Books.cs]: make file name the same (DbBooks, not Books)
and usually use DbBook, DbLibrary and so on; but table names Books, Libraries..

File: HW4/Models/Books.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        //public const string TABLE_NAME = "Books";
```
Comment (vorobevamv) [inline HW4/Models/Books.cs]: use it or don't write

File: HW4/Models/Orders.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿
+namespace HW4.Models
+{
+    public class DbOrders
+    {
+
```
Comment (vorobevamv) [inline HW4/Models/Orders.cs]: remove empty line

File: HW4/Models/Orders.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        public Guid Id { get; set; }
+
+        public Guid UserId { get; set; }
+        public DbUsers User { get; set; }
+
+        public Guid BookId  { get; set; }
+        public DbBooks Book { get; set; }
```
Comment (vorobevamv) [inline HW4/Models/Orders.cs]: write first Models properties (~columns), then properties-connections
```
public Guid Id { get; set; }
public Guid UserId { get; set; }
public Guid BookId { get; set; }

public DbUsers Users { get; set; }
public DbBooks Books { get; set; }
```

File: HW4/Models/Users.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿namespace HW4.Models
+{
+    public class DbUsers
+    {
+        //public const string TABLE_NAME = "Users";
+
+        public Guid Id { get; set; }
+        public string Name { get; set; }
+        public ICollection<DbOrders> Orders { get; set; }
```
Comment (vorobevamv) [inline HW4/Models/Users.cs]: add empty line

File: HW4/Data/LibContext.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Npgsql.EntityFrameworkCore;
+using HW4.Models;
+
+namespace HW4.Data
+{
+    public class LibContext : DbContext
```
Comment (vorobevamv) [inline HW4/Data/LibContext.cs]: add "Db" to name (LibDbContext)

File: HW4/Data/LibContext.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        protected override void OnConfiguring(DbContextOptionsBuilder options)
+        {
+            options.UseNpgsql("Host=localhost;Port=5432;Database=Lib;Username=postgres;Password=poo");
```
Comment (vorobevamv) [inline HW4/Data/LibContext.cs]: you use PostgreSQL? not SQLServer?
it's OK, but in project we use SQLServer, so to participate it project, you would have to install it

File: HW4/Data/LibContext.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        protected override void OnModelCreating(ModelBuilder builder)
```
Comment (vorobevamv) [inline HW4/Data/LibContext.cs]: why here (usually is written with DbModels)?
why don't use method Configure (from IEntityConfiguration)?

File: HW4/Data/LibContext.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            builder.Entity<DbBooks>()
+                .HasOne(o=>o.Library)
+                .WithMany(o=>o.Books)
+                .HasForeignKey(o=>o.LibraryId);
```
Comment (vorobevamv) [inline HW4/Data/LibContext.cs]: what about "HasKey"? You should set primary key!

File: HW4/Data/LibContext.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+                .HasOne(o => o.User)
+                .WithMany(o => o.Orders)
+                .HasForeignKey(o => o.UserId);
+        }
+
+        public DbSet<DbLibraries> Libraries { get; set; }
+        public DbSet<DbBooks> Books { get; set; }
+        public DbSet<DbUsers> Users { get; set; }
+        public DbSet<DbOrders> Orders { get; set; }
+
```
Comment (vorobevamv) [inline HW4/Data/LibContext.cs]: remove empty line

File: HW4/Menues/DB/BookController.cs [hunk 11]
```diff
@@ -0,0 +1,100 @@
+﻿using ConsoleManagement;
+using HW4.Data;
+using HW4.Models;
+
+namespace HW4.Menues.DB
+{
+    internal class BookController : IController, IOption
```
Comment (vorobevamv) [inline HW4/Menues/DB/BookController.cs]: it's Console Application (not web); what for write Contollers?
Rename it, please

File: HW4/Menues/DB/UserController.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            context.Users.Add(ReceiveInputForCreate());
+        }
+
+        private DbUsers ReceiveInputForCreate()
+        {
+            ConsoleHelper.WriteService("Enter user name");
+            string name = Console.ReadLine();
+            return new DbUsers
+            {
+                Id=Guid.NewGuid(),
```
Comment (vorobevamv) [inline HW4/Menues/DB/UserController.cs]: use spaces
Id = Guid.NewGuid();

File: HW4/Menues/DB/UserController.cs [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            string name = ReceiveInputForDelete();
+            var user = context.Users.FirstOrDefault(u => u.Name == name); 
+
+            if (user == null)
+            {
+                ConsoleHelper.WriteError("This user does'nt exist");
+                return;
+            }
+
+            context.Users.Remove(user);
```
Comment (vorobevamv) [inline HW4/Menues/DB/UserController.cs]: does it works?
what about "SaveChanges"?)

File: HW4/Menues/DB/UserController.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        private string ReceiveInputForDelete()
```
Comment (vorobevamv) [inline HW4/Menues/DB/UserController.cs]: don't mix private and public methods; usually order is like this:
class...Name
properties/fields
private methods
constructor
public methods

File: HW4/Menues/DB/UserController.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            ConsoleHelper.WriteService("Enter name for delete");
+            return Console.ReadLine();
+        }
+
+        public void Read(LibContext context)
+        {
+            var users = context.Users.ToList();
+
+            foreach (var user in users)
+                ConsoleHelper.WriteResult(String.Join(',', user.Id, user.Name));
```
Comment (vorobevamv) [inline HW4/Menues/DB/UserController.cs]: use parentheses

File: HW4/Menues/DB/UserController.cs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+            (string nameForUpdate, string newName) = ReceiveInputForUpdate();
+            var user = context.Users.FirstOrDefault(u => u.Name == nameForUpdate);
+            
+            if(user == null)
+            {
+                ConsoleHelper.WriteError("This user does'nt exist");
+                return;
+            }
+
+            user.Name = newName;
```
Comment (vorobevamv) [inline HW4/Menues/DB/UserController.cs]: SaveChanges!

File: HW4/Menues/Fibonacci/Fibonacci.cs [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace HW4.Menues.Fibonacci
+{
+    internal class Fibonacci : IOption
+    {
+        public string OptionName { get => "Calculate number"; }
+
+
+
+        public void Run()
```
Comment (vorobevamv) [inline HW4/Menues/Fibonacci/Fibonacci.cs:9]: 8+9 - remove; not more than one empty line to separate

File: HW4/Migrations/_20221231195900_InitialCreate.cs [hunk 18]
```diff
@@ -0,0 +1,80 @@
+﻿using HW4.Data;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace HW4.Migrations
+{
+
```
Comment (vorobevamv) [inline HW4/Migrations/_20221231195900_InitialCreate.cs]: remove empty line

File: HW4/Migrations/_20221231195900_InitialCreate.cs [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    [DbContext(typeof(LibContext))]
+    [Migration("20230101222800_InitialCreate")]
+    public class InitialCreate : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name:"Libraries",
+                columns:table => new
```
Comment (vorobevamv) [inline HW4/Migrations/_20221231195900_InitialCreate.cs]: add spaces:
name: "Libraries",
columns: table => new

File: HW4/Migrations/_20221231195900_InitialCreate.cs [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    Address = table.Column<string>(nullable: true),
+                },
+                constraints: table => 
+                {
+                    table.PrimaryKey("PK_Library", l => l.Id);
+                }
+                );
```
Comment (vorobevamv) [inline HW4/Migrations/_20221231195900_InitialCreate.cs]: it can be in one line
});

File: HW4/Migrations/_20221231195900_InitialCreate.cs [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),        
+                    Title = table.Column<string>(nullable:true),
+                    LibraryId = table.Column<Guid>(nullable: false)
+                },
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_Books", b => b.Id);
+                    table.ForeignKey("FK_Library_Books", b => b.LibraryId, "Libraries", "Id");
```
Comment (vorobevamv) [inline HW4/Migrations/_20221231195900_InitialCreate.cs]: it's not good idea to set foreign key in Initial Migration. Foreign key should be added, when needed (there is a lot of data in table, search takes too much time). There is no reason to use foreign key with small tables => search will be the same, but creating, updating and deleting will be slower

File: HW4/Program.cs [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+namespace HW4
+{
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+
+            using (LibContext context = new LibContext())
+            {
+                context.Database.Migrate();
```
Comment (vorobevamv) [inline HW4/Program.cs:17]: close using DbContext after this line;
what if user just want to count Fibonachchi? - why should it be inside using?
you need DbContext for migration and for CRUD, not for anything else

File: HW4/Data/LibContext.cs (hunk 6 above)
Comment (TheDioBrando) [reply HW4/Data/LibContext.cs]: Done

File: HW4/Menues/DB/BookController.cs (hunk 11 above)
Comment (TheDioBrando) [reply HW4/Menues/DB/BookController.cs]: Is "command" correct naming?

File: HW4/Data/LibContext.cs (hunk 7 above)
Comment (TheDioBrando) [reply HW4/Data/LibContext.cs]: Change to SQL server. 
I have 2019 version. Is it okay?

File: HW4/Data/LibContext.cs (hunk 8 above)
Comment (TheDioBrando) [reply HW4/Data/LibContext.cs]: Replace configure to models folders

File: HW4/Data/LibContext.cs (hunk 9 above)
Comment (TheDioBrando) [reply HW4/Data/LibContext.cs]: Add HasKey method

File: HW4/Data/LibContext.cs (hunk 10 above)
Comment (TheDioBrando) [reply HW4/Data/LibContext.cs]: Done

File: HW4/Menues/DB/UserController.cs (hunk 12 above)
Comment (TheDioBrando) [reply HW4/Menues/DB/UserController.cs]: Done

File: HW4/Menues/DB/UserController.cs (hunk 13 above)
Comment (TheDioBrando) [reply HW4/Menues/DB/UserController.cs]: I put SaveChanges to DbMenu class SelectCommand method to follow DRY pattern.
It works fine.
Is it correct?

File: HW4/Menues/DB/UserController.cs (hunk 14 above)
Comment (TheDioBrando) [reply HW4/Menues/DB/UserController.cs]: Done

File: HW4/Menues/DB/UserController.cs (hunk 15 above)
Comment (TheDioBrando) [reply HW4/Menues/DB/UserController.cs]: Parentheses have been added

File: HW4/Menues/DB/UserController.cs (hunk 16 above)
Comment (TheDioBrando) [reply HW4/Menues/DB/UserController.cs]: Put SaveChanges to DbMenu class SelectCommand method

File: HW4/Menues/Fibonacci/Fibonacci.cs (hunk 17 above)
Comment (TheDioBrando) [reply HW4/Menues/Fibonacci/Fibonacci.cs:9]: Done

File: HW4/Migrations/_20221231195900_InitialCreate.cs (hunk 18 above)
Comment (TheDioBrando) [reply HW4/Migrations/_20221231195900_InitialCreate.cs]: Done

File: HW4/Migrations/_20221231195900_InitialCreate.cs (hunk 19 above)
Comment (TheDioBrando) [reply HW4/Migrations/_20221231195900_InitialCreate.cs]: Done

File: HW4/Migrations/_20221231195900_InitialCreate.cs (hunk 20 above)
Comment (TheDioBrando) [reply HW4/Migrations/_20221231195900_InitialCreate.cs]: Fixed

File: HW4/Migrations/_20221231195900_InitialCreate.cs (hunk 21 above)
Comment (TheDioBrando) [reply HW4/Migrations/_20221231195900_InitialCreate.cs]: Removed FKs

File: HW4/Models/Books.cs (hunk 1 above)
Comment (TheDioBrando) [reply HW4/Models/Books.cs]: Renamed

File: HW4/Models/Books.cs (hunk 2 above)
Comment (TheDioBrando) [reply HW4/Models/Books.cs]: Removed

File: HW4/Models/Orders.cs (hunk 3 above)
Comment (TheDioBrando) [reply HW4/Models/Orders.cs]: Done

File: HW4/Models/Orders.cs (hunk 4 above)
Comment (TheDioBrando) [reply HW4/Models/Orders.cs]: Fixed

File: HW4/Models/Users.cs (hunk 5 above)
Comment (TheDioBrando) [reply HW4/Models/Users.cs]: Added

File: HW4/Program.cs (hunk 22 above)
Comment (TheDioBrando) [reply HW4/Program.cs:17]: Replace using closing bracket.
Replace context to DbMenu class

**Pull request merged**